R/grid_search.R

```diff
@@ -91,147 +94,175 @@ scan_beta_date <- function(
   }
 
   if (is.null(pars_obs)) {
-    pars_obs <- list(phi_general = 0.95,
-                     k_general = 2,
-                     phi_ICU = 0.95,
-                     k_ICU = 2,
-                     phi_death = 926 / 1019,
-                     k_death = 2,
-                     exp_noise = 1e6)
+    pars_obs <- list(
+      phi_general = 0.95,
+      k_general = 2,
+      phi_ICU = 0.95,
+      k_ICU = 2,
+      phi_death = 926 / 1019,
+      k_death = 2,
+      exp_noise = 1e6
+    )
   }
-  
+
   #
   # Multi-core futures with furrr (parallel purrr)
   #
   ## Particle filter outputs, extracting log-likelihoods
   pf_run_ll <- furrr::future_pmap_dbl(
     .l = param_grid, .f = beta_date_particle_filter,
-    sircovid_model = sircovid_model, model_params = model_params, data = data, 
+    sircovid_model = sircovid_model, model_params = model_params, data = data,
     pars_obs = pars_obs, n_particles = n_particles,
     forecast_days = 0, save_particles = FALSE, return = "ll"
   )
-  
+
   ## Construct a matrix with start_date as columns, and beta as rows
   ## order of return is set by order passed to expand.grid, above
   ## Returned column-major (down columns of varying beta) - set byrow = FALSE
   mat_log_ll <- matrix(
-    pf_run_ll, 
+    pf_run_ll,
     nrow = length(beta_1D),
     ncol = length(date_list),
-    byrow = FALSE)
+    byrow = FALSE
+  )
 
   # Exponentiate elements and normalise to 1 to get probabilities
   prob_matrix <- exp(mat_log_ll)
-  renorm_mat_LL <- prob_matrix/sum(prob_matrix)
-  
+  ## Check if the edges of the matrix in each dimension are close
+  ## enough to 0.
+  ## The first and last rows and the first and last columns
+  close_enough <- (all(prob_matrix[1, ] < tolerance) &
```

Comment:
I think it's worth generalising this, as although I'm not sure if we will extend past 2D these are the cases where this check is most useful as it becomes difficult to visualise the results. I would move this to a function that looks something like this:
```
# from https://stackoverflow.com/a/14502298
index_array <- function(x, dim, value, drop = FALSE) { 
  # Create list representing arguments supplied to [
  # bquote() creates an object corresponding to a missing argument
  indices <- rep(list(bquote()), length(dim(x)))
  indices[[dim]] <- value

  # Generate the call to [
  call <- as.call(c(
    list(as.name("["), quote(x)),
    indices,
    list(drop = drop)))
  # Print it, just to make it easier to see what's going on
  print(call)

  # Finally, evaluate it
  eval(call)
}

zero_boundary <- function(array, tolerance) {
  if (class(array) == "matrix") {
    close_enough <- (all(prob_matrix[1, ] < tolerance) &
    all(prob_matrix[, 1] < tolerance) &
    all(prob_matrix[nrow(prob_matrix), ] < tolerance) &
    all(prob_matrix[, ncol(prob_matrix)] < tolerance)
  } else {
    close_enough <- 1
    for (dimension in 1:dim(array)) {
      close_enough <- (all(index_array(array, dimension, 1) < tolerance &
       all(index_array(array, dimension, dim(array)[dimension]) < tolerance)
    }
  }
 close_enough
}